tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,208 @@
+        message = (
+            f"LLM update for {project_name} failed at submission time: "
+            + response.json()["message"]
+        )
+        LOGGER.error(message)
+        raise RuntimeError(message)
+    operation_id = response.content.decode("latin1")
+
+    # An operation id of '-1' is returned when LLM is already up to date
+    if operation_id != "-1":
+        operation = Operation.from_resource_id(tamr_client, operation_id)
+        operation.wait()
+    return None
+
+
+def poll_llm_status(
+    match_client: Client, *, project_name: str, num_tries: int = 10, wait_sec: int = 1
+) -> bool:
+    """
+    Check if LLM is queryable. Try up to num_tries times at 1s (or user-specified) interval.
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        num_tries: max number of times to poll endpoint, default 10
+        wait_sec: number of seconds to wait between tries, default 1
+    Returns:
+        bool indicating whether project is queryable
+    """
+
+    url = f"/api/v1/projects/{project_name}:isQueryable"
+    counter = 0
+
+    # Poll endpoint to until project is queryable or num_tries reached
+    while counter < num_tries:
+
+        response = match_client.get(url)
+        queryable = response.content == b"true"
+
+        if queryable:
+            break
+
+        counter += 1
+        time.sleep(wait_sec)  # call api at wait_sec interval if project isn't yet queryable
+
+    return queryable
+
+
+def llm_query(
+    match_client: Client,
+    *,
+    project_name: str,
+    records: Union[JsonDict, List[JsonDict]],
+    type: str,
+    pKey: Optional[str] = None,
+    batch_size: Optional[int] = None,
+    min_match_prob: Optional[float] = None,
+    max_num_matches: Optional[int] = None,
+) -> Dict[int, List[JsonDict]]:
+    """
+    Find the best matching clusters or records for each supplied record. Returns empty list if
+    response is null.
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        records: record or list of records to match
+        type: one of "records" or  "clusters" -- whether to pull record or cluster matches
+        pKey: a primary key for the data; if supplied, this must be a field in the input records
+        batch_size: split input into this batch size for LLM calls (e.g. to prevent network
+            timeouts), default None sends a single LLM call with all records
+        min_match_prob: if set, only matches with probability above minimum will be returned,
+            default None
+        max_num_matches: if set, at most max_num_matches will be returned for each input record in
+            records, default None
+    Returns:
+        Dict keyed by integers (indices of inputs), or by pKey if pKey is supplied, with value a
+            list containing matcched data
+    Raises:
```

Comment:
Docstring formatting and typo
```suggestion
            records, default None
    
    Returns:
        Dict keyed by integers (indices of inputs), or by pKey if pKey is supplied, with value a
            list containing matched data
    
    Raises:
```